pkgs/native_assets_cli/lib/src/locking/locking.dart

```diff
@@ -0,0 +1,120 @@
+///
+/// If provided, the [timeout] parameter determines how long to retry before
+/// giving up. If the [timeout] is exceeded a [TimeoutException] is thrown.
+///
+/// If provided, the [logger] streams information on the the locking status, and
+/// also streams error messages.
+Future<T> runUnderDirectoryLock<T>(
+  Directory directory,
+  Future<T> Function() callback, {
+  Duration? timeout,
+  Logger? logger,
+}) async {
+  const lockFileName = '.lock';
+  final lockFile = File.fromUri(directory.uri.resolve(lockFileName));
+  return _runUnderFileLock(
+    lockFile,
+    callback,
+    timeout: timeout,
+    logger: logger,
+  );
+}
+
+/// Run [callback] with this Dart process having exclusive access to [file].
+///
+/// Note multiple isolates and isolate groups in the same Dart process share
+/// locks, so these will be able to enter the exclusive section simultanously.
+///
+/// If provided, the [timeout] parameter determines how long to retry before
+/// giving up. If the [timeout] is exceeded a [TimeoutException] is thrown.
+///
+/// If provided, the [logger] streams information on the the locking status,
+/// and also streams error messages.
+Future<T> _runUnderFileLock<T>(
+  File file,
+  Future<T> Function() callback, {
+  Duration? timeout,
+  Logger? logger,
+}) async {
+  await _ensureLockfileExists(file, logger);
+
+  var timedOut = false;
+  Timer? timeoutTimer;
+  if (timeout != null) {
+    timeoutTimer = Timer(timeout, () {
+      timedOut = true;
+    });
+  }
+
+  final randomAccessFile = await file.open(mode: FileMode.write);
+  var locked = false;
+  var printed = false;
+  while (!locked && !timedOut) {
+    try {
+      await randomAccessFile.lock(FileLock.exclusive);
+      locked = true;
+    } on FileSystemException {
+      if (!printed) {
+        logger?.finer(
+          'Waiting to be able to obtain lock of directory: ${file.path}.',
+        );
+        printed = true;
+      }
+      await Future<void>.delayed(const Duration(milliseconds: 50));
+    }
+  }
+
+  if (locked) {
+    timeoutTimer?.cancel();
+    try {
+      if (!Platform.isWindows) {
+        // On Windows, a file can't be written to once we have a lock on it.
+        // https://github.com/dart-lang/sdk/issues/56378
+        await file.writeAsString(
+          'Last acquired by ${Platform.resolvedExecutable} '
+          'running ${Platform.script} on ${DateTime.now()}.',
+        );
+      }
+      return await callback();
+    } finally {
+      await randomAccessFile.unlock();
+    }
+  }
+  assert(timedOut);
+  final message = 'Could not acquire the lock to ${file.path}.';
+  logger?.severe(message);
+  throw TimeoutException(message, timeout);
+}
+
+Future<void> _ensureLockfileExists(
+  File file,
+  Logger? logger,
+) async {
+  while (true) {
+    try {
+      if (!await file.exists()) {
+        // This code could race with itself in another process.
+        await file.create(recursive: true, exclusive: true);
```

Comment:
Why `exclusive: true`? We just need it to be there. (One can use `File.create(exclusive: true)` to coordinate exclusivity between processes, but this is not what we rely on here - we rely on advisory locks which are taken later)

Also this code here is swallowing **all** `FileSystemException`s and continue to loop. This is problematic - for example if there's a permission error (i.e. the process doesn't have write access to the folder) then this may be an infinite loop.

The code will not be safe for racing deletions anyway (i.e. we check if it exists if not create it, thinking it's there - but right after that someone in another terminal could `rm -rf` it)

So overall I think this can be simplified to just
```
if (!await file.exists()) await file.create(recursive: true);
```

and since it's only a one liner, I'd not have a method for it, but just handle on the callsite.